pkg/datapath/iptables/ipset/ipset.go

```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ipset
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+	"net/netip"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/cilium/cilium/pkg/datapath/tables"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/ip"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/statedb/reconciler"
+)
+
+const (
+	CiliumNodeIpsetV4 = "cilium_node_set_v4"
+	CiliumNodeIpsetV6 = "cilium_node_set_v6"
+
+	IpsetFamilyV4 = "inet"
+	IpsetFamilyV6 = "inet6"
+)
+
+// Manager handles the kernel IP sets configuration
+type Manager interface {
+	AddToNodeIpset(nodeIP net.IP)
+	RemoveFromNodeIpset(nodeIP net.IP)
+}
+
+type manager struct {
+	logger  logrus.FieldLogger
+	enabled bool
+
+	db    *statedb.DB
+	table statedb.RWTable[*tables.Ipset]
+
+	ipset *ipset
+}
+
+// AddToNodeIpset adds an IP address to the ipset for cluster nodes. It creates
+// the ipset if it doesn't already exist and doesn't error if either the ipset
+// or the IP already exist.
+func (m *manager) AddToNodeIpset(nodeIP net.IP) {
+	if !m.enabled {
+		return
+	}
+
+	addr, ok := ip.AddrFromIP(nodeIP)
+	if !ok {
+		m.logger.WithField(logfields.IPAddr, nodeIP).Error("unable to convert to netip.Addr")
+		return
+	}
+
+	name := CiliumNodeIpsetV4
+	family := IpsetFamilyV4
+	if addr.Is6() {
+		name = CiliumNodeIpsetV6
+		family = IpsetFamilyV6
+	}
+
+	txn := m.db.WriteTxn(m.table)
+	defer txn.Abort()
+
+	var (
+		obj   *tables.Ipset
+		found bool
+	)
+
+	obj, _, found = m.table.First(txn, tables.IpsetNameIndex.Query(name))
+	if !found {
+		obj = &tables.Ipset{
+			Name:   name,
+			Family: family,
+			Addrs:  sets.New(addr),
+			Status: reconciler.StatusPending(),
+		}
+	} else {
+		obj.Addrs[addr] = sets.Empty{}
+		obj.Status = reconciler.StatusPending()
+	}
+	_, _, _ = m.table.Insert(txn, obj)
+	txn.Commit()
+}
+
+// RemoveFromBodeIpset removes an IP address from the ipset for cluster nodes.
+func (m *manager) RemoveFromNodeIpset(nodeIP net.IP) {
+	if !m.enabled {
+		return
+	}
+
+	addr, ok := ip.AddrFromIP(nodeIP)
+	if !ok {
+		m.logger.WithField(logfields.IPAddr, nodeIP).Error("unable to convert to netip.Addr")
+		return
+	}
+
+	name := CiliumNodeIpsetV4
+	if addr.Is6() {
+		name = CiliumNodeIpsetV6
+	}
+
+	txn := m.db.WriteTxn(m.table)
+	defer txn.Abort()
+
+	obj, _, found := m.table.First(txn, tables.IpsetNameIndex.Query(name))
+	if !found || !obj.Addrs.Has(addr) {
+		return
+	}
+	delete(obj.Addrs, addr)
```

Comment:
Same here, you cannot mutate the objects returned by queries.

Reply:
Unless of course it's returned by value and not be reference (e.g. `Ipset` and not `*Ipset`, though fields themselves need still a copy)

Reply:
I kept the `*Ipset` type since `reconciler.New` requires a `comparable`.